cloud/openstack/os_security_group_rule.py

```diff
@@ -0,0 +1,214 @@
+        - ID of Security group to link (exclusive with remote_ip_prefix)
+      required: false
+   ethertype:
+      description:
+        - Must be IPv4 or IPv6, and addresses represented in CIDR must
+          match the ingress or egress rules. Not all providers support IPv6.
+      choices: ['IPv4', 'IPv6']
+      default: IPv4
+   direction:
+     description:
+        - The direction in which the security group rule is applied. Not
+          all providers support egress.
+     choices: ['egress', 'ingress']
+     default: ingress
+   state:
+     description:
+       - Should the resource be present or absent.
+     choices: [present, absent]
+     default: present
+requirements: ["shade"]
+'''
+# TODO(mordred): add ethertype and direction
+
+EXAMPLES = '''
+# Create a security group rule
+- os_security_group_rule:
+    cloud: mordred
+    security_group: foo
+    protocol: tcp
+    port_range_min: 80
+    port_range_max: 80
+    remote_ip_prefix: 0.0.0.0/0
+'''
+
+
+def _find_matching_rule(module, secgroup):
+    """
+    Find a rule in the group that matches the module parameters.
+    :returns: The matching rule dict, or None if no matches.
+    """
+    protocol = module.params['protocol']
+    port_range_min = module.params['port_range_min']
+    port_range_max = module.params['port_range_max']
+    remote_ip_prefix = module.params['remote_ip_prefix']
+    ethertype = module.params['ethertype']
+    direction = module.params['direction']
+
+    for rule in secgroup['security_group_rules']:
+        # No port, or -1, will be returned from shade as None
+        if rule['port_range_min'] is None:
+            rule_port_range_min = -1
+        else:
+            rule_port_range_min = int(rule['port_range_min'])
+
+        if rule['port_range_max'] is None:
+            rule_port_range_max = -1
+        else:
+            rule_port_range_max = int(rule['port_range_max'])
+
+
+        if (protocol == rule['protocol']
+                and port_range_min == rule_port_range_min
+                and port_range_max == rule_port_range_max
+                and remote_ip_prefix == rule['remote_ip_prefix']
+                and ethertype == rule['ethertype']
+                and direction == rule['direction']):
+            return rule
+    return None
+
+
+def _system_state_change(module, secgroup):
+    state = module.params['state']
+    if secgroup:
+        rule_exists = _find_matching_rule(module, secgroup)
+    else:
+        return False
+
+    if state == 'present' and not rule_exists:
+        return True
+    if state == 'absent' and rule_exists:
+        return True
+    return False
+
+
+def main():
+    argument_spec = openstack_full_argument_spec(
+        security_group   = dict(required=True),
+        protocol         = dict(default='tcp',
+                                choices=['tcp', 'udp', 'icmp']),
+        port_range_min   = dict(required=True, type='int'),
+        port_range_max   = dict(required=True, type='int'),
```

Comment:
With Neutron port ranges are optional, with a default of Null, which would span all ports. As such this module should not require port_range_min/max and default to None.